src/pcie_errors.c

```
@@ -0,0 +1,779 @@
+      sstrncpy(pcie_config.access_dir, PCIE_DEFAULT_SYSFSDIR,
+               sizeof(pcie_config.access_dir));
+  } else {
+    /* use proc */
+    pcie_fops.list_devices = pcie_list_devices_proc;
+    pcie_fops.open = pcie_open_proc;
+    if (pcie_config.access_dir[0] == '\0')
+      sstrncpy(pcie_config.access_dir, PCIE_DEFAULT_PROCDIR,
+               sizeof(pcie_config.access_dir));
+  }
+  /* Common functions */
+  pcie_fops.close = pcie_close;
+  pcie_fops.read = pcie_read;
+}
+
+static int pcie_plugin_config(oconfig_item_t *ci) {
+
+  for (int i = 0; i < ci->children_num; i++) {
+    oconfig_item_t *child = ci->children + i;
+    int status = 0;
+
+    if (strcasecmp("Source", child->key) == 0) {
+      if ((child->values_num != 1) ||
+          (child->values[0].type != OCONFIG_TYPE_STRING)) {
+        status = -1;
+      } else if (strcasecmp("proc", child->values[0].value.string) == 0) {
+        pcie_config.use_sysfs = 0;
+      } else if (strcasecmp("sysfs", child->values[0].value.string) != 0) {
+        ERROR(PCIE_ERRORS_PLUGIN ": Allowed sources are 'proc' or 'sysfs'.");
+        status = -1;
+      }
+    } else if (strcasecmp("AccessDir", child->key) == 0) {
+      status = cf_util_get_string_buffer(child, pcie_config.access_dir,
+                                         sizeof(pcie_config.access_dir));
+    } else if (strcasecmp("ReportMasked", child->key) == 0) {
+      status = cf_util_get_boolean(child, &pcie_config.notif_masked);
+    } else if (strcasecmp("PersistentNotifications", child->key) == 0) {
+      status = cf_util_get_boolean(child, &pcie_config.persistent);
+    } else {
+      ERROR(PCIE_ERRORS_PLUGIN ": Invalid configuration option \"%s\".",
+            child->key);
+      pcie_config.config_error = 1;
+      break;
+    }
+
+    if (status) {
+      ERROR(PCIE_ERRORS_PLUGIN ": Invalid configuration parameter \"%s\".",
+            child->key);
+      pcie_config.config_error = 1;
+      break;
+    }
+  }
+
+  return 0;
```

please return an error on error.

---

The idea here is to keep collectd running even if one plugin fails. If plugin config fails collectd stops loading. So error is stored in `pcie_config.config_error` and plugin returns error later on init, while the rest of the plugins can continue running until someone fixes the configuration and restart. Does it sound reasonable?

---

I do understand your reasoning. However, if a plugin is misconfigured collectd shouldn't even start and the configuration should be fixed, making the whole system more predicatable. Configuration can then also be checked with collectd -t.

---

We have to put this as a rule in developers documentation / style guide, and do some work to unify existing plugins behaviour. Yesterday I worked on `snmp` plugin and may say what it can start with wrong configuration (when `Data` block referenced from `Host` is not found, for example).

Maybe it's a time to do this?

---

Yes, good idea.